### DISCUSSION TITLE
Should the query history be persistent or is it per-session?

### DISCUSSION THREAD

**Discussion by curio77** [tconbeer/harlequin#631] [Q&A, Troubleshooting, Support]
Query history seems to be per-session, but this is not clarified anywhere in the documentation.  If it being per-session and not persistent is intentional and not a bug, I think it would be beneficial if it could optionally be persisted across sessions, say, via a local query history file (à la `.bash_history`) enabled by a config option.

**Comment by tconbeer** [comment]
We try to persist it across sessions for the same connection. I actually just changed that way Harlequin computes a hash key from the connection so if you upgraded recently that could have wiped out your history.

**Comment by curio77** [comment]
Thank you for the quick response.

Hmm, always getting “Harlequin could not load the Query History.”/“Your Query History has not yet been loaded. Please wait a moment and try again.” after starting Harlequin.

And as for the same-connection aspect: For example, my “~/.duckdbrc” attaches a local MariaDB database, so this appears automatically under “Databases” when I launch Harlequin.  Yet the history is always empty initially, also after USEing that DB.  Is that to be expected?

**Reply by tconbeer** [reply]
That notification appears when you try to load the history before the connection to the database has been established. If you wait and try again it should work?

**Comment by curio77** [comment]
No, it doesn't.  So to avoid any misunderstandings, I run Harlequin without any options/local config, so it uses its DuckDB backend.  My `~/.duckdbrc` attaches a (MariaDB/MySQL) database I frequently use.  At no point after restarting Harlequin I get to see commands from previous sessions (not right after start-up, not after `USE attached_database`, never).  The only thing that persists is the contents of the Query Editor from when the previous session was quit, but nothing in the Query History (aforementioned “not yet loaded” message until I run a query, then it is populated with new queries).  Is that the expected behavior?

**Reply by tconbeer** [reply]
no, definitely not. Can you open a [bug report](https://github.com/tconbeer/harlequin/issues/new?assignees=&labels=&projects=&template=bug_report.md&title=)?
